### ISSUE TITLE
Why model.evaluate() and model.predict() have different accuracy results

### ISSUE THREAD

**Issue by YJZh** [tflearn/tflearn#507]
I have the same problems as https://github.com/tflearn/tflearn/issues/223 mentioned. But mine is related to multi-class classification.

        self.metric = self.model.evaluate(self.case.X_test, self.case.y_test, batch_size=self.BATCH_SIZE)
        print('metric = ',self.metric)
        
        self.y_pre = [self.model.predict([i]) for idx,i in enumerate( self.case.X_test)]
        self.y_pre = np.squeeze(self.y_pre, 1)
        print(np.shape(self.case.y_test),np.shape(self.y_pre))
        
        import sklearn.metrics as metrics
        print("\tTest fscore:\t  ",str(metrics.f1_score(np.argmax(self.case.y_test,1), np.argmax(self.y_pre,1), average='weighted')))
        test_true = np.argmax(self.case.y_test,1)
        test_pred = np.argmax(self.y_pre,1)
        print ("Precision: {}%".format(100*metrics.precision_score(test_true, test_pred, average="weighted")))
        print ("Recall: {}%".format(100*metrics.recall_score(test_true, test_pred, average="weighted")))
        print ("f1_score: {}%".format(100*metrics.f1_score(test_true, test_pred, average="weighted")))

and the result displays:
metric = 0.8242
Precision: 40.49679836186083%
Recall: 45.77536477774007%
f1_score: 39.44194888503528%

I tried the same dataset as that used at https://github.com/guillaume-chevalier/LSTM-Human-Activity-Recognition

Thank you

**Comment by aymericdamien** [comment]
Does the results differ a lot? If not, It may be a problem that at every prediction, results are rounded (to the X decimal), so it may explain.

**Comment by xypzk** [comment]
I'm seeing the same problem with multi-class classification. Accuracy difference between evaluate and predict is about 31% ! Here are the nos:

Validation accuracy using evaluate: 0.7559999995231629
Validation accuracy using predict: 0.3425

Python script that demonstrates the problem:
`import csv
import numpy as np
from sklearn.model_selection import train_test_split
import tensorflow as tf
import tflearn

def buildModel(X, Y, numLabels):
    tf.reset_default_graph()
    # Build a deep neural network with 2 fully connected layers and an
    #  output softmax layer
    input_layer = tflearn.input_data(shape=[None, 4])
    dense1 = tflearn.fully_connected(input_layer, 128, activation='tanh',
                                     regularizer='L2', weight_decay=0.001)
    dropout1 = tflearn.dropout(dense1, 0.8)
    dense2 = tflearn.fully_connected(dropout1, 128, activation='tanh',
                                     regularizer='L2', weight_decay=0.001)
    dropout2 = tflearn.dropout(dense2, 0.8)
    softmax = tflearn.fully_connected(dropout2, numLabels, activation='softmax')
    
    # Regression using SGD with learning rate decay and Top-3 accuracy
    sgd = tflearn.SGD(learning_rate=0.1, lr_decay=0.96, decay_step=1000)
    top_k = tflearn.metrics.Top_k(3)
    net = tflearn.regression(softmax, optimizer=sgd, metric=top_k,
                             loss='categorical_crossentropy')
    
    # Train the model
    model = tflearn.DNN(net, tensorboard_verbose=0)
    model.fit(X, Y, n_epoch=15, show_metric=True, run_id="dnnModel")
              
    return model
    
if __name__ == "__main__":
    with open('./testData.csv', 'r') as f:
        readData = list(csv.reader(f, delimiter=','))
    
    inpData = np.array(readData).astype(np.float)
    numLabels = np.max(inpData[:, -1]).astype(int)
    labelsForMdl = np.zeros((inpData.shape[0], numLabels), dtype = int)
    
    # Store the label in one-hot vector form.
    for idx, curData in enumerate(inpData):
        labelsForMdl[idx][np.int(curData[-1]) - 1] = 1
        
    xTrain, xVal, yTrain, yVal = train_test_split(inpData[:,0:4], labelsForMdl,
                                                  test_size=0.20, random_state=41)
    dnnModel = buildModel(xTrain, yTrain, numLabels)
    print("\nValidation accuracy using evaluate: {}".format(
            dnnModel.evaluate(xVal, yVal)[0]))
    
    valPred = np.array(dnnModel.predict(xVal))
    correctRatio = np.mean(np.equal(np.argmax(valPred, axis=1), 
                                    np.argmax(yVal, axis=1)).astype(int))
    print("Validation accuracy using predict: {}".format(correctRatio))`

Data set to reproduce the problem is attached.
[testData.csv.zip](https://github.com/tflearn/tflearn/files/691728/testData.csv.zip)

I'll post an update if I make progress with debugging the problem.

Thanks

**Comment by xypzk** [comment]
Here is the python script and data-set as a zip file in case code from the previous comment is unusable.
[script+testData.zip](https://github.com/tflearn/tflearn/files/691731/script.testData.zip)

**Comment by hosna-sattar** [comment]
try to remove dropout at the test time and see if you get the same result as model.evaluate()
